scene/gui/color_picker_shape.cpp

```diff
@@ -0,0 +1,474 @@
+	if (mev.is_valid() && shape->changing_color) {
+		Vector2 uv = shape->get_square_uv(Rect2(Vector2(), shape->sv_square->get_size()), mev->get_position());
+		color_picker->s = uv.x;
+		color_picker->v = 1.0 - uv.y;
+
+		shape->apply_picker_color();
+		shape->emit_color_changed_if_not_deferred();
+	}
+}
+
+void ColorPickerShapeRectangle::_hue_slider_input(const Ref<InputEvent> &p_event, int64_t p_shape) {
+	ColorPickerShapeRectangle *shape = reinterpret_cast<ColorPickerShapeRectangle *>(p_shape);
+	ColorPicker *color_picker = shape->color_picker;
+	const float h = color_picker->h;
+	Control *hue_slider = shape->hue_slider;
+
+	Ref<InputEventMouseButton> bev = p_event;
+	if (bev.is_valid() && bev->get_button_index() == MouseButton::LEFT) {
+		if (bev->is_pressed()) {
+			shape->changing_color = true;
+			float y = bev->get_position().y;
+			float height = hue_slider->get_size().y;
+			y = CLAMP(y, 0, height);
+			color_picker->h = y / height;
+		} else {
+			shape->changing_color = false;
+		}
+		shape->apply_picker_color();
+
+		if (!bev->is_pressed()) {
+			shape->add_recent_preset();
+			shape->emit_color_changed();
+		} else {
+			shape->emit_color_changed_if_not_deferred();
+		}
+	}
+
+	Ref<InputEventMouseMotion> mev = p_event;
+	if (mev.is_valid() && shape->changing_color) {
+		float y = mev->get_position().y;
+		float height = hue_slider->get_size().y;
+		y = CLAMP(y, 0, height);
+		color_picker->h = y / height;
+
+		shape->apply_picker_color();
+		shape->emit_color_changed_if_not_deferred();
+	}
+}
+
+void ColorPickerShapeRectangle::_sv_square_draw(int64_t p_shape) {
+	ColorPickerShapeRectangle *shape = reinterpret_cast<ColorPickerShapeRectangle *>(p_shape);
+	Control *sv_square = shape->sv_square;
+	shape->draw_sv_square(sv_square, Rect2(Vector2(), sv_square->get_size()));
+}
+
+void ColorPickerShapeRectangle::_hue_slider_draw(int64_t p_shape) {
+	ColorPickerShapeRectangle *shape = reinterpret_cast<ColorPickerShapeRectangle *>(p_shape);
+	Control *hue_slider = shape->hue_slider;
+
+	const ColorPicker::ThemeCache &theme_cache = shape->color_picker->theme_cache;
+	const float h = shape->color_picker->h;
+
+	const Vector2 size = hue_slider->get_size();
+	hue_slider->draw_texture_rect(theme_cache.color_hue, Rect2(0, 0, -size.y, size.x), false, Color(1, 1, 1), true);
+
+	int y = size.y * h;
+	Color color = Color::from_hsv(h, 1, 1);
+	hue_slider->draw_line(Vector2(0, y), Vector2(size.x, y), color.inverted());
+}
+
+void ColorPickerShapeRectangle::update_theme() {
+	const ColorPicker::ThemeCache &theme_cache = color_picker->theme_cache;
+	sv_square->set_custom_minimum_size(Size2(theme_cache.sv_width, theme_cache.sv_height));
+	hue_slider->set_custom_minimum_size(Size2(theme_cache.h_width, 0));
+}
+
+ColorPickerShapeRectangle::ColorPickerShapeRectangle(ColorPicker *p_color_picker) :
+		ColorPickerShape(p_color_picker) {
+	int64_t self = reinterpret_cast<int64_t>(this); // Hack to allow connecting non-Object methods.
```

Comment:
I don't like this hack, but I didn't have better idea. Alternative is making the shapes Object/RefCounted, but that would make them heavier than necessary (and each ColorPicker has a copy). It would make the code cleaner though.